src/categorical_algebra/DataMigration.jl

```diff
@@ -0,0 +1,159 @@
+""" Data Migration functors
+"""
+module DataMigration
+
+export Σ, Functor
+
+using ...Theories
+using ...Theories: CatDesc, AttrDesc, ob, hom, dom, codom
+using ...CSetDataStructures
+using ..FinSets, ..CSets, ..Limits, ..FreeDiagrams
+using ...Graphs
+using ...Present
+
+import ...CategoricalAlgebra.FreeDiagrams: FreeDiagram
+
+""" Functor
+
+A functor ``F: \\mathcal{C} \\to \\mathcal{D}`` consists of a map of objects and a 
+map of homomorphisms from a domain category ``\\mathcal{C}``  to a codomain
+category ``\\mathcal{D}``.
+"""
+struct Functor{Schema, Ob, Hom}
+  FOb::Dict{Ob, Ob}
+  FHom::Dict{Hom, Hom}
+  dom::Presentation{Schema, Symbol}
+  codom::Presentation{Schema, Symbol}
+end
+
+""" Inv(F::Functor)
+
+For a functor ``F: \\mathcal{C} \to \\mathcal{D}``, 
+returns a function which when given an object (resp. morphism)
+of ``\\mathcal{D}``, returns an array of objects (resp. morphisms) 
+of ``\\mathcal{C}`` in the preimage.
+"""
+function Inv(F::Functor{Schema, Ob, Hom}) where {Schema, Ob, Hom}
+  y::Union{Ob, Hom} -> begin
+    y isa Ob ? D = F.FOb : D = F.FHom
+    [k for (k,v) in D if v == y]
+  end
+end
+
+
+""" add_hom!(d::FreeDiagram{Ob, Hom}, src_ob::Ob, tgt_ob::Ob, hom::Hom)
+
+Adds a hom to `d` from the vertex with object `src_ob` to the vertex with object `tgt_ob`.
+"""
+function add_hom!(d::FreeDiagram, src_ob, tgt_ob, hom) 
+  src_idx = first(incident(d, src_ob, :ob))
+  tgt_idx = first(incident(d, tgt_ob, :ob))
+  return add_edge!(d, src_idx, tgt_idx, hom = hom)
+  end
+
+""" FreeDiagram(pres::Presentation{Schema, Symbol})
+
+Returns a `FreeDiagram` whose objects are the generating objects of `pres` and 
+whose homs are the generating homs of `pres`.
+"""
+function FreeDiagram(pres::Presentation{Schema, Symbol}) where {Schema}
+  obs = Array{FreeSchema.Ob}(generators(pres, :Ob))
+  homs = Array{FreeSchema.Hom}(generators(pres, :Hom))
+  doms = map(h -> generator_index(pres, nameof(dom(h))), homs)
+  codoms = map(h -> generator_index(pres, nameof(codom(h))), homs)
+  return FreeDiagram(obs, collect(zip(homs, doms, codoms)))
+end
+
+
+""" Σ(F::Functor)
+
+which takes a ``\\mathcal{C}``-Set ``X: \\mathcal{C} \\to \\mathsf{Set}`` 
+and produces the pushforward ``\\Sigma_F(X): \\mathcal{D} \\to \\mathsf{Set}``.
+
+``\\mathcal{D}`` must be a free schema with no cycles. 
+"""
+function Σ(F::Functor{Schema, Ob, Hom}) where {Schema, Ob, Hom}
```

Comment:
again, might be best to make a struct SigmaFunctor and then overload calling

Reply:
also, would be nice if this function was split up a little

Reply:
and if there was documentation of the overall strategy

Reply:
I like the idea of `Σ(F::AbstractFunctor)::SigmaMigration <: AbstractFunctor` also, we always been ASCII names in case your text editor is broken and you are writing in an old version of `vi` with no unicode support.